### ISSUE TITLE
Cannot import pytables on windows, python 3.8

### ISSUE THREAD

**Issue by aamster** [PyTables/PyTables#933]
- python 3.8
- pytables 3.7.0
- windows
- installed via `pip install tables`

```
import tables
...\site-packages\tables\__init__.py", line 45, in <module>
    from .utilsextension import get_hdf5_version as _get_hdf5_version
ImportError: DLL load failed while importing utilsextension: The specified module could not be found
```

This seems to indicate that something is not getting registered properly on windows, python 3.8. Why is it unable to find `utilsextension`?

**Comment by avalentino** [comment]
Hi @aamster thanks for reporting.
Can you please confirm that tables have been installed via wheels?

Also could you please post the list of files in `...\site-packages\tables`?

**Comment by aamster** [comment]
Hi @avalentino 

`ls ...\site-packages\tables`
```
Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----         1/3/2022  11:06 AM                misc
d-----         1/3/2022  11:06 AM                nodes
d-----         1/3/2022  11:06 AM                scripts
d-----         1/3/2022  11:06 AM                tests
d-----         1/3/2022  11:06 AM                __pycache__
-a----         1/3/2022  11:06 AM          36021 array.py
-a----         1/3/2022  11:06 AM          42702 atom.py
-a----         1/3/2022  11:06 AM          26710 attributeset.py
-a----         1/3/2022  11:06 AM         112128 blosc.dll
-a----         1/3/2022  11:06 AM         187904 bzip2.dll
-a----         1/3/2022  11:06 AM          10612 carray.py
-a----         1/3/2022  11:06 AM          16280 conditions.py
-a----         1/3/2022  11:06 AM          22574 definitions.pxd
-a----         1/3/2022  11:06 AM          39101 description.py
-a----         1/3/2022  11:06 AM           9702 earray.py
-a----         1/3/2022  11:06 AM          11164 exceptions.py
-a----         1/3/2022  11:06 AM          28661 expression.py
-a----         1/3/2022  11:06 AM         108572 file.py
-a----         1/3/2022  11:06 AM          16586 filters.py
-a----         1/3/2022  11:06 AM          13321 flavor.py
-a----         1/3/2022  11:06 AM          48807 group.py
-a----         1/3/2022  11:06 AM        3503616 hdf5.dll
-a----         1/3/2022  11:06 AM        1689289 hdf5extension.c
-a----         1/3/2022  11:06 AM         285696 hdf5extension.cp38-win_amd64.pyd
-a----         1/3/2022  11:06 AM            958 hdf5extension.pxd
-a----         1/3/2022  11:06 AM          76172 hdf5extension.pyx
-a----         1/3/2022  11:06 AM          16817 idxutils.py
-a----         1/3/2022  11:06 AM          92466 index.py
-a----         1/3/2022  11:06 AM           6053 indexes.py
-a----         1/3/2022  11:06 AM        1242613 indexesextension.c
-a----         1/3/2022  11:06 AM         121856 indexesextension.cp38-win_amd64.pyd
-a----         1/3/2022  11:06 AM          50554 indexesextension.pyx
-a----         1/3/2022  11:06 AM          27656 leaf.py
-a----         1/3/2022  11:06 AM          13388 link.py
-a----         1/3/2022  11:06 AM         565201 linkextension.c
-a----         1/3/2022  11:06 AM          74752 linkextension.cp38-win_amd64.pyd
-a----         1/3/2022  11:06 AM           8372 linkextension.pyx
-a----         1/3/2022  11:06 AM         754874 lrucacheextension.c
-a----         1/3/2022  11:06 AM          97792 lrucacheextension.cp38-win_amd64.pyd
-a----         1/3/2022  11:06 AM           2441 lrucacheextension.pxd
-a----         1/3/2022  11:06 AM          22271 lrucacheextension.pyx
-a----         1/3/2022  11:06 AM          33071 node.py
-a----         1/3/2022  11:06 AM          15154 parameters.py
-a----         1/3/2022  11:06 AM           6859 path.py
-a----         1/3/2022  11:06 AM           2315 registry.py
-a----         1/3/2022  11:06 AM            628 req_versions.py
-a----         1/3/2022  11:06 AM         147199 table.py
-a----         1/3/2022  11:06 AM        1220594 tableextension.c
-a----         1/3/2022  11:06 AM         191488 tableextension.cp38-win_amd64.pyd
-a----         1/3/2022  11:06 AM          61238 tableextension.pyx
-a----         1/3/2022  11:06 AM           4038 undoredo.py
-a----         1/3/2022  11:06 AM           5265 unimplemented.py
-a----         1/3/2022  11:06 AM          14086 utils.py
-a----         1/3/2022  11:06 AM        1119698 utilsextension.c
-a----         1/3/2022  11:06 AM         198656 utilsextension.cp38-win_amd64.pyd
-a----         1/3/2022  11:06 AM            756 utilsextension.pxd
-a----         1/3/2022  11:06 AM          46321 utilsextension.pyx
-a----         1/3/2022  11:06 AM          32848 vlarray.py
-a----         1/3/2022  11:06 AM          84992 zlib.dll
-a----         1/3/2022  11:06 AM         131284 _comp_bzip2.c
-a----         1/3/2022  11:06 AM          24576 _comp_bzip2.cp38-win_amd64.pyd
-a----         1/3/2022  11:06 AM            394 _comp_bzip2.pyx
-a----         1/3/2022  11:06 AM         131140 _comp_lzo.c
-a----         1/3/2022  11:06 AM          28160 _comp_lzo.cp38-win_amd64.pyd
-a----         1/3/2022  11:06 AM            388 _comp_lzo.pyx
-a----         1/3/2022  11:06 AM           6401 __init__.py
```
> Can you please confirm that tables have been installed via wheels?
What do you mean by this? How can I check this?

**Comment by aamster** [comment]
FYI this issue specifically occurs with Pytables 3.7.0. It does not occur with 3.6.1

**Comment by avalentino** [comment]
@aamster it is really strange, from the list you posted it seems that the `utilsextension` for Python 3.8 is there: `utilsextension.cp38-win_amd64.pyd`.
... and also libraries it depends on: `blosc.dll`, `bzip2.dll`.
Probably only `zstd.dll` is missing but it depends on how the package has been built.

>> Can you please confirm that tables have been installed via wheels?
>
> What do you mean by this? How can I check this?

I would like to know if the package has been built locally form sources or if the `pip` exploited the wheel package `tables-3.7.0-cp38-cp38-win_amd64.whl` for the installation.
The output of the `pip install tables` command could help.

**Comment by aamster** [comment]
```
Collecting tables<4.0.0,>=3.6.0
  Downloading tables-3.7.0-cp38-cp38-win_amd64.whl (6.9 MB)
```

**Comment by avalentino** [comment]
Dear @aamster I have updated the recipe to produce binary wheels on windows platforms.
Could you please give a try to the windows wheel included in the [archive of artifacts](https://github.com/avalentino/PyTables/suites/4840840617/artifacts/137247322) of the GHA run at: https://github.com/avalentino/PyTables/actions/runs/1659070310?

**Comment by cgohlke** [comment]
I can't reproduce this with the official Python.org distribution. This is likely a miniconda issue related to the `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` magic added to `__init__.py`.  Adding a `from . import utilsextension` right after `os.add_dll_directory(libs_dir)` and before restoring `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` might work.

**Comment by avalentino** [comment]
Sorry @cgohlke probably I'm missing something, your hint is for @aamster or for PyYbables. In the second case, could you please be more clear about what we need to do?

**Comment by cgohlke** [comment]
The `tables-3.7.0` wheels for Windows on PyPI have the following code added to `__init__.py`:
```Python
""""""  # start delvewheel patch
def _delvewheel_init_patch_0_0_17():
    import os
    import sys
    libs_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir, 'tables.libs'))
    if sys.version_info[:2] >= (3, 8):
        if os.path.exists(os.path.join(sys.base_prefix, 'conda-meta')):
            # backup the state of the environment variable CONDA_DLL_SEARCH_MODIFICATION_ENABLE
            conda_dll_search_modification_enable = os.environ.get("CONDA_DLL_SEARCH_MODIFICATION_ENABLE")
            os.environ['CONDA_DLL_SEARCH_MODIFICATION_ENABLE']='1'

        os.add_dll_directory(libs_dir)

        if os.path.exists(os.path.join(sys.base_prefix, 'conda-meta')):
            # restore the state of the environment variable CONDA_DLL_SEARCH_MODIFICATION_ENABLE
            if conda_dll_search_modification_enable is None:
                os.environ.pop("CONDA_DLL_SEARCH_MODIFICATION_ENABLE", None)
            else:
                os.environ["CONDA_DLL_SEARCH_MODIFICATION_ENABLE"] = conda_dll_search_modification_enable
    else:
        from ctypes import WinDLL
        with open(os.path.join(libs_dir, '.load-order-tables-3.7.0')) as file:
            load_order = file.read().split()
        for lib in load_order:
            WinDLL(os.path.join(libs_dir, lib))


_delvewheel_init_patch_0_0_17()
del _delvewheel_init_patch_0_0_17
# end delvewheel patch
```

Apparently this does not work as expected using miniconda:
```
Python 3.8.10 (default, May 19 2021, 13:12:57) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import tables
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "...\tables\__init__.py", line 45, in <module>
    from .utilsextension import get_hdf5_version as _get_hdf5_version
ImportError: DLL load failed while importing utilsextension: The specified module could not be found.
```

Adding `from . import utilsextension` right after `os.add_dll_directory(libs_dir)` works (the `bzip2.dll` not found isue is already fixed in master):
```Python
""""""  # start delvewheel patch
def _delvewheel_init_patch_0_0_17():
    import os
    import sys
    libs_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir, 'tables.libs'))
    if sys.version_info[:2] >= (3, 8):
        if os.path.exists(os.path.join(sys.base_prefix, 'conda-meta')):
            # backup the state of the environment variable CONDA_DLL_SEARCH_MODIFICATION_ENABLE
            conda_dll_search_modification_enable = os.environ.get("CONDA_DLL_SEARCH_MODIFICATION_ENABLE")
            os.environ['CONDA_DLL_SEARCH_MODIFICATION_ENABLE']='1'

        os.add_dll_directory(libs_dir)
        from . import utilsextension  # <-- added

        if os.path.exists(os.path.join(sys.base_prefix, 'conda-meta')):
            # restore the state of the environment variable CONDA_DLL_SEARCH_MODIFICATION_ENABLE
            if conda_dll_search_modification_enable is None:
                os.environ.pop("CONDA_DLL_SEARCH_MODIFICATION_ENABLE", None)
            else:
                os.environ["CONDA_DLL_SEARCH_MODIFICATION_ENABLE"] = conda_dll_search_modification_enable
    else:
        from ctypes import WinDLL
        with open(os.path.join(libs_dir, '.load-order-tables-3.7.0')) as file:
            load_order = file.read().split()
        for lib in load_order:
            WinDLL(os.path.join(libs_dir, lib))


_delvewheel_init_patch_0_0_17()
del _delvewheel_init_patch_0_0_17
# end delvewheel patch
```
```
Python 3.8.10 (default, May 19 2021, 13:12:57) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import tables
>>> tables.test()
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
PyTables version:    3.7.0
HDF5 version:        1.12.0
NumPy version:       1.22.0
Numexpr version:     2.8.1 (not using Intel's VML/MKL)
Zlib version:        1.2.11 (in Python interpreter)
Blosc version:       1.21.0 (2020-12-22)
Blosc compressors:   blosclz (2.3.0), lz4 (1.9.3), lz4hc (1.9.3), zlib (1.2.11), zstd (1.4.5)
Blosc filters:       shuffle, bitshuffle
Python version:      3.8.10 (default, May 19 2021, 13:12:57) [MSC v.1916 64 bit (AMD64)]
Platform:            Windows-10-10.0.19044-SP0
Byte-ordering:       little
Detected cores:      12
Default encoding:    utf-8
Default FS encoding: utf-8
Default locale:      (en_US, cp1252)
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
Performing only a light (yet comprehensive) subset of the test suite.
If you want a more complete test, try passing the --heavy flag to this script
(or set the 'heavy' parameter in case you are using tables.test() call).
The whole suite will take more than 4 hours to complete on a relatively
modern CPU and around 512 MB of main memory.
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
tables\filters.py:305: FiltersWarning: compression library ``bzip2`` is not available; using ``zlib`` instead
  warnings.warn("compression library ``%s`` is not available; "
tables\filters.py:305: FiltersWarning: compression library ``lzo`` is not available; using ``zlib`` instead
  warnings.warn("compression library ``%s`` is not available; "
<snip>
```

**Comment by avalentino** [comment]
Thanks @cgohlke, it is clear now.
I think that there is not too much that we can do in this case, apart for providing an updated conda package maybe.
I have prepared [PR #76](https://github.com/conda-forge/pytables-feedstock/pull/76) for the coda-forge [feedstock](https://github.com/conda-forge/pytables-feedstock) but there seems to be some problem with PyPy on aarch64 and ppc64le architectures.

**Comment by xmatthias** [comment]
I had identified the same issue with conda/miniconda in the "windows wheels pr" (https://github.com/PyTables/PyTables/pull/926#issuecomment-994437720).
I don't have enough knowledge about conda/miniconda to determine why conda is not loading the dll correctly.

If it's a problem with the added code in `__init__.py` - this should be reported to delvewheel / cibuildwheel.

**Comment by aufdenkampe** [comment]
The issue with installing from Conda might be related to the fact that the latest version of pytables on Conda-forge is 3.6.1. https://anaconda.org/conda-forge/pytables

Is there a plan to update the Conda-forge recipe to include the 3.7.0 release?

**Comment by avalentino** [comment]
@aufdenkampe a PR is already available: https://github.com/conda-forge/pytables-feedstock/pull/76

**Comment by aufdenkampe** [comment]
That's great to know! Thanks!

**Comment by hyjforesight** [comment]
same issue here https://github.com/theislab/scanpy/issues/2108

**Comment by rly** [comment]
I have the same issue with Python 3.9. To recap this thread, the latest pytables (3.7.0) cannot be imported using pip (`pip install tables`) in a conda environment with Python 3.8 or 3.9 on windows due to the failed DLL load error above. One workaround is to install pytables using conda and conda-forge (`conda install -c conda-forge tables`). However, as far as I understand, for packages that depend on pytables and are only pip-installable, a user cannot install these packages in a conda environment with Python 3.8 or 3.9 on windows. Another workaround is to set the environment variable `CONDA_DLL_SEARCH_MODIFICATION_ENABLE=1`, but it would be ideal if pytables and its downstream packages could just work out of the box without needing this fix.

This issue has been fixed in Python 3.9.9+. But not everyone can use Python 3.9.9+. Would it be possible to push a bugfix in pytables?

One solution based on above comments and https://github.com/adang1345/delvewheel/blob/3a0d77d38e9c81c03617a65f9893b5ef04504207/delvewheel/wheel_repair.py#L43-L54 is to set the environment variable `CONDA_DLL_SEARCH_MODIFICATION_ENABLE=1` immediately prior to calling `from .utilsextension import get_hdf5_version as _get_hdf5_version` in `tables/__init__.py` and then reset the env var afterward.

For example:
Current code:
https://github.com/PyTables/PyTables/blob/6a0cd0008fc8a149628c413ee79af06e44e217b7/tables/__init__.py#L10-L11

New:
```python
import os as _os
import sys as _sys
import platform as _platform
_conda_workaround = (_platform.system().lower() == 'windows' and _sys.version_info[:2] >= (3, 8) and
        _sys.version_info[:3] < (3, 9, 9) and _os.path.exists(_os.path.join(_sys.base_prefix, 'conda-meta')))
if _conda_workaround:
    # backup the state of the environment variable CONDA_DLL_SEARCH_MODIFICATION_ENABLE
    _conda_dll_search_modification_enable = _os.environ.get('CONDA_DLL_SEARCH_MODIFICATION_ENABLE')
    _os.environ['CONDA_DLL_SEARCH_MODIFICATION_ENABLE'] = '1'

# Necessary imports to get versions stored on the cython extension
from .utilsextension import get_hdf5_version as _get_hdf5_version

if _conda_workaround:
    # restore the state of the environment variable CONDA_DLL_SEARCH_MODIFICATION_ENABLE
    if _conda_dll_search_modification_enable is None:
        _os.environ.pop('CONDA_DLL_SEARCH_MODIFICATION_ENABLE', None)
    else:
        _os.environ['CONDA_DLL_SEARCH_MODIFICATION_ENABLE'] = _conda_dll_search_modification_enable
    del _conda_dll_search_modification_enable
del _conda_workaround
```

I admit this code is ugly, but it works. Perhaps instead, it would work to add similar C code here:
https://github.com/PyTables/PyTables/blob/6a0cd0008fc8a149628c413ee79af06e44e217b7/src/utils.c#L22-L43

See also the open issue in conda: https://github.com/conda/conda/issues/10897

**Comment by avalentino** [comment]
Dear @rly, thanks for your summary.
Of course we are open to consider PR fixing this issue but please consider that path the moment we do not have plans to make a new releasee in the near future.
It can be discussed of course.

By the way it is not too clear to me the following statement

> However, as far as I understand, for packages that depend on pytables and are only pip-installable, a user cannot install these packages in a conda environment with Python 3.8 or 3.9 on windows.

I agree that pip cannot trigger a conda install but if you are already in a Conda environment what prevents you to prepare it by installing potables?

**Comment by rly** [comment]
> I agree that pip cannot trigger a conda install but if you are already in a Conda environment what prevents you to prepare it by installing potables?

Sorry, my comment was not exactly correct. You can install pytables there but it requires another step. Example: package my_library requires pytables. my_library is available only on PyPI. If a user is running a conda environment with Python 3.8/3.9 on Windows and wants to install my_library or any application that requires my_library, they would have to do the following:
```bash
pip install my_library  # does not result in an error on install but `import tables` results in DLL load error
pip uninstall tables
conda install -c conda-forge pytables  # but generally not recommended to install with conda after pip
```
or
```bash
conda install -c conda-forge pytables  # user must know that this is a prerequisite for their system
pip install my_library
```

**Comment by avalentino** [comment]
yes, that is what I meant.

Also, one could include potables in its environment file to automatically have it installed when the environment is created.

I understand that it is annoying but not a blocking issue.

By the way, as I said, we are open to PRs to fix this issue.

**Comment by avalentino** [comment]
As far as I can understand this is fixed in PyTables v3.8.
Please re-open if it is not the case
